py/persipubsub/subscriber.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+"""Receive messages persistently from the queue."""
+
+import contextlib
+import datetime
+import pathlib
+import time
+from typing import Any, Iterator, Optional, Union
+
+import icontract
+import lmdb  # pylint: disable=unused-import
+
+import persipubsub.queue
+
+# pylint: disable=protected-access
+
+
+class Subscriber:
+    """
+    Handle receiving messages stored in the queue.
+
+    :ivar identifier: subscriber ID
+    :vartype pub_id: str
+    :ivar queue: from which messages are received
+    :vartype queue: persipubsub.queue.Queue
+    """
+
+    def __init__(self) -> None:
+        """Initialize class object."""
+        self.identifier = None  # type: Optional[str]
+        self.queue = None  # type: Optional[persipubsub.queue._Queue]
+
+    def init(self,
+             identifier: str,
+             path: Union[pathlib.Path, str],
+             env: Optional[lmdb.Environment] = None) -> None:
+        """
+        Initialize.
+
+        :param identifier: unique subscriber id
+        :param path: path to the queue
+        :param env: open lmdb environment
+        """
+        self.identifier = identifier
+        assert isinstance(self.identifier, str)
+        self.queue = persipubsub.queue._Queue()  # pylint: disable=protected-access
+        self.queue.init(path=path, env=env)
+        assert isinstance(self.queue, persipubsub.queue._Queue)
+
+    def __enter__(self) -> 'Subscriber':
+        """Enter the context and give the sub prepared in the constructor."""
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        """Leave the context."""
+
+    @icontract.require(lambda timeout: timeout > 0)
+    @icontract.require(lambda retries: retries > 0)
+    @contextlib.contextmanager
+    def receive(self, timeout: int = 60,
+                retries: int = 10) -> Iterator[Optional[bytes]]:
+        """
+        Receive messages from the queue.
+
+        :param timeout: time waiting for a message. If none arrived until the
+            timeout then None will be returned. (secs)
+        :param retries: number of tries to check if a msg arrived in the queue
+        """
+        msg = None
+        end = int(datetime.datetime.utcnow().timestamp()) + timeout
+        assert isinstance(self.queue, persipubsub.queue._Queue)
+        assert isinstance(self.identifier, str)
+        try:
+            while int(datetime.datetime.utcnow().timestamp()) <= end:
+                msg = self.queue.front(identifier=self.identifier)
+                if msg is not None:
+                    break
+                time.sleep(timeout / retries)
+            yield msg
+        finally:
+            pass
+
+        if msg is not None:
+            self._pop()
```

Comment:
Are you sure the correct message will be popped in a multi-thread/multi-process setting? Since `self.queue.front` and `self.queue.pop` operate on different transaction, I think this will result in buggy behavior. @radomsak could we please have a look at this today together?